ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/BlsToExecutionOperationPool.java

```diff
@@ -0,0 +1,198 @@
+    final Collection<SignedBlsToExecutionChange> sortedViableOperations = operations.values();
+    final List<SignedBlsToExecutionChange> selected = new ArrayList<>();
+    for (final SignedBlsToExecutionChange item : sortedViableOperations) {
+      if (!filter.test(item)) {
+        continue;
+      }
+      final int validatorIndex = getValidatorIndex(item);
+      if (operationValidator.validateForBlockInclusion(stateAtBlockSlot, item).isEmpty()) {
+        selected.add(item);
+        includedItemConsumer.accept(item);
+        if (selected.size() == schema.getMaxLength()) {
+          break;
+        }
+      } else {
+        // The item is no longer valid to be included in a block so remove it from the pool.
+        operations.remove(validatorIndex);
+      }
+    }
+    return schema.createFromElements(selected);
+  }
+
+  @Override
+  public SafeFuture<InternalValidationResult> addLocal(final SignedBlsToExecutionChange item) {
+    final int validatorIndex = getValidatorIndex(item);
+    if (operations.containsKey(validatorIndex)) {
+      return SafeFuture.completedFuture(rejectForDuplicatedMessage(validatorIndex))
+          .thenPeek(result -> validationReasonCounter.labels(result.code().toString()).inc());
+    }
+    return add(item, false);
+  }
+
+  @Override
+  public SafeFuture<InternalValidationResult> addRemote(final SignedBlsToExecutionChange item) {
+    final int validatorIndex = getValidatorIndex(item);
+    if (operations.containsKey(validatorIndex)) {
+      return SafeFuture.completedFuture(rejectForDuplicatedMessage(validatorIndex))
+          .thenPeek(result -> validationReasonCounter.labels(result.code().toString()).inc());
+    }
+    return add(item, true);
+  }
+
+  @Override
+  public void addAll(final SszCollection<SignedBlsToExecutionChange> items) {
+    items.forEach(
+        item -> {
+          final int validatorIndex = getValidatorIndex(item);
+          operations.putIfAbsent(validatorIndex, item);
+        });
+  }
+
+  @Override
+  public void removeAll(final SszCollection<SignedBlsToExecutionChange> items) {
+    items.forEach(
+        item -> {
+          final int validatorIndex = getValidatorIndex(item);
+          operations.remove(validatorIndex);
+        });
+  }
+
+  @Override
+  public Set<SignedBlsToExecutionChange> getAll() {
+    return Set.copyOf(operations.values());
+  }
+
+  @Override
+  public int size() {
+    return operations.size();
+  }
+
+  private SafeFuture<InternalValidationResult> add(
+      SignedBlsToExecutionChange item, boolean fromNetwork) {
+    final int validatorIndex = getValidatorIndex(item);
+    return operationValidator
+        .validateForGossip(item)
+        .thenApply(
+            result -> {
+              validationReasonCounter.labels(result.code().toString()).inc();
+              if (result.code().equals(ValidationResultCode.ACCEPT)
+                  || result.code().equals(ValidationResultCode.SAVE_FOR_FUTURE)) {
```

Comment:
I believe for bls_to_execution_change operations we should never see SAVE_FOR_FUTURE (if we do it is a bug). We could remove it from this if clause now that we have a specific pool for bls ops.

Reply:
I've just left this because it was already in the queue, but i can remove it?

Reply:
I don't think our bls validator will ever return SAVE_FOR_FUTURE.

And the spec for bls to execution change does not specify any path for it. So I'd say we are better with you that.

Not a biggie but removing code we don't need is always a win!

Reply:
okies, i just was planning on being 'the same' as the simple pool...